tests/models/components/test_mixture_density_network.py

```diff
@@ -0,0 +1,101 @@
+        mixture = NormalMixture(log_pi, mu, sigma)
+
+        # Test batch_shape
+        assert mixture.batch_shape == torch.Size(batch_shape)
+
+        # Test sampling
+        sample = mixture.sample()
+        assert sample.shape == torch.Size(batch_shape)
+
+        # Test log_prob
+        log_prob = mixture.log_prob(sample)
+        assert log_prob.shape == torch.Size(batch_shape)
+
+        sample_shape = (10, 5)
+        samples = mixture.sample(sample_shape)
+        assert samples.shape == torch.Size(sample_shape + batch_shape)
+        assert mixture.log_prob(samples).shape == sample_shape + batch_shape
+
+        # Test rsample
+        rsample = mixture.rsample()
+        assert rsample.shape == torch.Size(batch_shape)
+
+        # Test consistency with individual normal components
+        components = [Normal(mu[..., i], sigma[..., i]) for i in range(num_components)]
+        mixture_log_prob = mixture.log_prob(sample)
+        component_log_probs = torch.stack([comp.log_prob(sample) for comp in components], dim=-1)
+        component_log_probs += log_pi
+        expected_log_prob = torch.logsumexp(component_log_probs, dim=-1)
+        assert torch.allclose(mixture_log_prob, expected_log_prob, atol=1e-5)
+
+    def test_normal_mixture_invalid_args(self):
+        # Test error handling for invalid arguments
+        with pytest.raises(AssertionError):
+            NormalMixture(torch.randn(3, 2), torch.randn(3, 3), torch.rand(3, 2).add_(0.1))
+
+
+class TestNormalMixtureDensityNetwork:
+    @pytest.mark.parametrize("in_features", [10, 20])
+    @pytest.mark.parametrize("out_features", [5, 8])
+    @pytest.mark.parametrize("num_components", [2, 3])
+    @pytest.mark.parametrize("batch_size", [1, 32])
+    def test_normal_mixture_density_network(self, in_features, out_features, num_components, batch_size):
+        # Create NormalMixtureDensityNetwork instance
+        network = NormalMixtureDensityNetwork(in_features, out_features, num_components)
+
+        # Create input tensor
+        x = torch.randn(batch_size, in_features)
+
+        # Forward pass
+        output = network(x)
+
+        # Check output type
+        assert isinstance(output, NormalMixture)
+
+        # Check output shapes
+        assert output.batch_shape == torch.Size([batch_size, out_features])
+        assert output.event_shape == torch.Size([])
+        assert output.log_pi.shape == (batch_size, out_features, num_components)
+        assert output.mu.shape == (batch_size, out_features, num_components)
+        assert output.sigma.shape == (batch_size, out_features, num_components)
+
+        # Check that sigma is positive
+        assert (output.sigma > 0).all()
+
+        # Check that log_pi is a valid log probability
+        assert torch.allclose(output.log_pi.exp().sum(dim=-1), torch.ones(batch_size, out_features))
+
+    def test_normal_mixture_density_network_gradients(self):
+        in_features, out_features, num_components = 10, 5, 3
```

Comment:
`@pytest.mark.parametrize`はこんな書き方もできたりする　これは複数の引数をセットで指定する例
```suggestion
    @pytest.mark.parametrize(
        ["in_features", "out_features", "num_components"],
        [
            [10, 5, 3],
        ]
    )
    def test_normal_mixture_density_network_gradients(
        self, 
        in_features: int, 
        out_features: int, 
        num_components: int,
    ):
```
テストケースを追加するときは例えば下みたいにすればもう1つ増やせて、合計2パターンについてテストできる
```
@pytest.mark.parametrize(
    ["in_features", "out_features", "num_components"],
    [
        [10, 5, 3],
        [9999999, 2147483648, 123456789],
    ]
)
```